### ISSUE TITLE
Move dynamically added styles into Shadow DOM

### ISSUE THREAD

**Issue by ghost** [GoogleWebComponents/google-chart#26]
It would be good if google-chart support the Table graph.

**Comment by SmokyBob** [comment]
+1

**Comment by SmokyBob** [comment]
Hi @sgomes, glad to help.
I took sometime thinking about what could be a solution that is both functional and the least difficult to manage.
Would be awesome to be able to directly get the styles from the visualization API with something like `google.visualization.getStyles()` or `google.visualization.Table.getStyles()` to get the URL and add the import in the shadow DOM from the component without using and observer... but for now observing the head element for changes seems the only viable way.
My2Cents

**Comment by SmokyBob** [comment]
Tried to set up the observer to get the CSS.
No problem identifying when the CSS element is added to the head, when the browser finish to load the stylesheet.... but I have no Idea how to get the CSS styles themselves to add them in the shadowroot...

[Here a gist](https://gist.github.com/SmokyBob/0a263be5e8cae094f0bd) with the code I've tried
CORS is not supported so the first piece of code doesn't work, the second waits for the `link` element to be loaded and try to search for one of the classes that are loaded by the API as you can find out in the devtools...

If anyone has any other Ideas, they are more than welcome

**Comment by sgomes** [comment]
So to add to this, as far as I can tell there's currently no way to implement support for the added styling.

It's easy enough to detect changes, as @SmokyBob mentions, but we then need a way to actually apply them as needed:
- adding a link inside the shadow DOM: this approach does not work, since links are ignored inside the shadow DOM (this was a surprise to me). Polymer's apparent support for `<link>` tags inside the element template is, I suspect, based on some clever static tricks; in any case, any dynamically inserted `<link>` tags get promptly ignored.
- adding a style tag with the content: as @SmokyBob mentioned, CORS is not enabled, so we can't make an XHR request to retrieve the styles.

Unless I'm missing something, the only solution for now seems to be to provide the markup, but no styling. Polymer 0.8 may give us a way of solving this, but for now it seems we're stuck :(

**Comment by philipwalton** [comment]
I think the `<google-chart>` element should package up some default styling in it's shadow DOM. That would be the "Web Component-y" way of doing things (rather than using global styles with the `/deep/`or `>>>` combinators). The default styling could be the same as the current table chart defaults, but I don't see any reason it has to be (it's kinda ugly).

If users want to override the default styles they have a few options:

1. Extend the `<google-chart>` element and set their own style overrides.
2. Compose the `<google-chart>` element inside of their own element and use the `::shadow` pseudo-element selector to override the default styles.
3. Set the `cssClassNames` chart option in combination with either 1) or 2) above to start fresh.

I'm happy to submit a PR with these changes as I want to get this working with the [`<google-analytics-chart>`](https://github.com/GoogleWebComponents/google-analytics/blob/v0.0.4/google-analytics-chart.html) element anyway.

@sgomes what do you think?

**Comment by SmokyBob** [comment]
@philipwalton IMO adding the default styling, even if a bit old and didn't age well is a good idea and compromise.
Using the current table styling just to keep the element behave the same as the Google Chart API and avoid causing possible confusion in the users that use the `google-chart` element.

For different, more modern, Material styles we could have something like a `materialTable` chart type to load the `google.Visualization.table` package but apply the new style?

**Comment by sgomes** [comment]
@philipwalton Thanks for helping out! I think it's a reasonable compromise, yes. I'm not thrilled about us keeping static CSS against a library with regular updates, but I still think it's a better approach than just throwing markup at users with no context.
As for what exactly the styling should look like, I don't think it's a big concern, as long as we have something and it's easily overridable (your suggested approach is perfect).

@SmokyBob: I'd rather stay away from calling anything "Material" unless it's based on the spec or has gone through the right reviews, so I think we're better just having the default styling. Again, it should be pretty easy for users to override styling, either outside of the component by using `::shadow`, or by extending it and modifying the styling.

**Comment by philipwalton** [comment]
Finally got around to working on this. Here's the PR: https://github.com/GoogleWebComponents/google-chart/pull/34

**Comment by SmokyBob** [comment]
Finally found a way to import the style sheets added by the Google Chart API.
tl;dr  PR: #38

Searching here and there found [this article](https://css-tricks.com/modular-future-web-components/) saying to use @import instead of the link tags... and works nicely.

**Comment by admwx7** [comment]
Any chance we'll get a resolution to this?  I'm in need of the org chart and it has this same issue.

**Comment by jongeho1** [comment]
Hi there,

I reported original related issue #74 and investigated this a bit at the time. A simple approach to resolving seems to be observe type + data changes, assure polymer settings dom is shadow, google vis + service loader api state and internalize the external css for each instance. Sample code:

````javascript

      // Lazily try to load table css for native shadow dom on type + data changes
      observers: [
          '_loadTableCss(type,data)'
      ],

      // State of table css for shadow dom
      _tableCssLoaded: false,

      // The type or data was changed
      // For shadowDom, we need to internalize css for type table
      _loadTableCss: function (type) {
          var loaded =  this._tableCssLoaded;
          var googReady = window.google != null && window.google.loader != null && window.google.visualization != null;
          var isShadowDom = window.Polymer!=null && window.Polymer.Settings.dom === "shadow";
          if (isShadowDom && !loaded && type == "table" && googReady) {
              this._tableCssLoaded = true;
              this.debounce("_loadTableCss", function(){
                  var url = window.google.loader.ServiceBase +
                          "/api/visualization/" +
                          window.google.visualization.Version +
                          "/" +
                          google.visualization.JSHash +
                          "/table+en.css";
                  var styleNode = document.createElement('style');
                  styleNode.innerHTML = "@import '" + url + "';";
                  this.shadowRoot.appendChild(styleNode);
              });
          }
      }
```

Some potential improvements would be to:

* if the css can be shared for all instances, then maintain a static variable for loaded state instead
* if  the locale needs to be parameterized, then sniff out or re-compute as necessary

Thanks

@jongeho1

**Comment by wesalvaro** [comment]
So, I believe we can do the following:
  - Add a [`scopeSubtree`](https://www.polymer-project.org/1.0/docs/devguide/styling#scope-subtree) container (perhaps we can reuse the chart container). This didn't seem necessary when testing.
    on-ready: `this.scopeSubtree(this.$.container, true);`
  - Add a callback to duplicate the CSS after packages are loaded.

Moving Callback:
  1. for-each `styleSheet` matching
    `head > link[id^="load-css-"][href^="https://www.gstatic.com/charts/"]`
  1. copy them to `this.shadowRoot` as a `link` or `style` node

This would mean that we have a duplicate of each sheet for each `google-chart` element, but should work generically and works for all dynamic style sheets (i.e. timeline, too).

Something like:

```js
listeners: {
  'loaded': '_checkForAndMoveStyles',
},
_checkForAndMoveStyles() {
  const styleElementId = 'dynamic-chart-styles';
  const chartStyleSelector = 'head > link[id^="load-css-"][href^="https://www.gstatic.com/charts/"]'

  const chartStyles = document.querySelectorAll(chartStyleSelector);
  if (!chartStyles) {
    return;
  }

  const imports = styleSheets.map((css) => {
     return css.getAttribute('href');
  }).join("';\n @import '");

  // In the case of multiple loads, an element already exists.
  let styleElement = this.shadowRoot.querySelector('#${styleElementId}');
  const existing = !!styleElement;
  if (!existing) {
    styleElement = document.createElement('style');
    styleElement.setAttribute('id', styleElementId)
  }
  styleElement.innerHTML = `@import '${imports}';`;
  if (!existing) {
    this.shadowRoot.appendChild(styleElement);
  }
},
```

**Comment by wesalvaro** [comment]
Just got around to testing this and it works well.
I think this is the way to go unless there are any other ideas or until something blows up. xD

**Comment by wesalvaro** [comment]
[Here](https://rawgit.com/GoogleWebComponents/google-chart/v2/demo/index.html) is a demo of it in action and a preview of the changes I had in mind.
